### PR TITLE
Add Window Title Editor plugin

### LINKED ISSUE #2017: Window Title Editor plugin
_When you want to pitch an idea:_
- _Consider going to the "discussions" forum first to discuss your idea; file it here only if it is fully detailed and ready for implementation_
- _Please provide a use case description; explain the problem you are trying to solve before describing the solution. This allows other users to think along and maybe improve on your solution._
- _Make sure the use case is generic enough that it will benefit other users as well. If it is highly tailored for your specific work-flow, chances are that no-one will work on it (and no-one will want to maintain it in working state over the years)._

**Is your feature request related to a problem? Please describe.**
_A clear and concise description of what the problem is. Ex. I'm always frustrated when [...]_
 
- Window titles start with the page path, which gets cut off in eg. the task bar.
- Windows belonging to different notebooks are hard to tell apart by title.
- The situation used to be better: the title used to start with the notebook name.
- Very well covered in the discussion.

**Describe the solution you'd like**
_A clear and concise description of what you want to happen._

There'd be a plugin - a _Window Title Editor plugin_, if you will - that'd allow the used to choose between a few ready-made templates, or create their own, like described in the issue.  `string.Templates.safe_substitute()` can be used to avoid `KeyErrors`.

**Describe alternatives you've considered**
Building this into Zim core as a hidden `notebook.zim` feature. That may be less hacky, but does introduce more maintenance burden.

**Additional context**

* There isn't a pretty API for this in [MainWindowExtension](https://introt.github.io/zim-desktop-wiki/docs/api/zim.gui.mainwindow.MainWindowExtension.html], but it should allow us to monkey patch [MainWindow._update_window_title(https://introt.github.io/zim-desktop-wiki/docs/api/zim.gui.mainwindow.MainWindow.html#_update_window_title)
    * For [PageWindow](https://introt.github.io/zim-desktop-wiki/docs/api/zim.gui.mainwindow.PageWindow.html), we can monkey patch the inherited `set_window` method
    * Reverting without restarts should be possible - just monkey patch the default methods back and hope Zim doesn't go bananas ;)
* I call Dib( Membrane)s on the implementation!
* Discussed in https://github.com/zim-desktop-wiki/zim-desktop-wiki/discussions/1660

<div type='discussions-op-text'>

<sup>Originally posted by **charles-esterbrook** September 28, 2021</sup>
## Title bar template preference

Zim would benefit from a new Interface preference that uses a Python templated string to let the user control the title bar. 

Fields could include:

- page name ("ZimIsGreat")
- page location ("NoteTaking:Applications")
- page path ("NoteTaking:Applications:ZimIsGreat")
- notebook name ("My Life")
- notebook path ("/home/me/Notes/MyLife")

Examples:

- `{notebook name} - Zim`
  - "My Life - Zim"
  - Zim 0.6x style
- `{page path} - {notebook name}`
  - "NoteTaking:Applications:ZimIsGreat - My Life"
  - Zim 0.74 style
- `{notebook name} - {page path}`
  - "My Life - NoteTaking:Applications:ZimIsGreat"
  - for #1339
- `Zim: {notebook name} in {notebook path}`
  - "Zim: My Life in /home/me/Notes/MyLife"
- `Page {page name} in {notebook name} - Zim`
  - "Page ZimIsGreat in My Life"
- etc.

Python's .format() method would be called on the template as in `titleBarTemplateStr.format(**fieldsDict)`, probably with `try:except KeyError:` protection.

## Motivation

When moving form Zim 0.6x to Zim 0.7x, I can no longer efficiently use the Windows 10 taskbar window previews to choose a notebook, due to changes in the title bar. Note that the window previews are shown when left clicking the Zim app icon in the taskbar. Right clicking shows a basic menu ("Zim Desktop Wiki", "Unpin from taskbar", "Close all windows").

You can see the taskbar-triggered window previews in the screenshot at this unrelated superuser issue:
https://superuser.com/questions/1136612/how-to-refresh-windows-10-taskbar-thumbnail-previews

Note that all the title bars are clipped with "...". You can only read the beginning of each title bar. This is key.

Suppose you have multiple Zim notebooks open and their names are:

- Artificial Intelligence
- TechNotes

Then the title bars in Zim 0.6x are:

- Artificial Intelligence - Zim
- TechNotes - Zim

And the Windows 10 window previews will show:

- Artificial Intel...
- TechNotes...

Which is easy to use (meaning choose from) because these are the names of the notebook, or at least their beginnings.

However, in Zim 0.7x, the title bars will start with the _current page name_ which changes quite frequently. Note only can the constant change of the beginning of the title bar make it more difficult to pick the notebook from a list, but two notebooks may have the identical name among the windows previews. Suppose you are looking at the page "Applications" in both notebooks. Then their window title bars are:

- Applications - Artificial Intelligence
- Applications - TechNotes

But the window list shows:

- Applications ...
- Applications ...

... that's _my_ situation. In #1339, another user wants to put the full path of the page in the title bar - after the notebook name. I personally don't want the page name let alone its full path, but I understand that people have different needs which may also vary by window manager. My needs could even change if I move from Windows to Mac or Linux.
</div>

### PR DESCRIPTION
**Pull request by introt**
Fixes #2017, closes #1660.

### CODE REVIEW

**Comment by mtill** [comment]
Hi,
can confirm this is working fine.
Thank you, introt, for this great extension!
BR
mtill

**Comment by introt** [comment]
@jaap-karssenberg could you consider including this plugin in 0.75? It does work via monkey-patching the title function, but imo is simple & has good enough test coverage for upstream inclusion.

**Comment by introt** [comment]
Fixed the manual typos - still need to decide on uniform quotation throughout the manual, but that's not relevant to this PR.

The failing check on windows seems to be unrelated; the plugin's own tests are ok.

```
2022-09-25T13:12:38.5426178Z testMainWindowApi (tests.windowtitleeditor.TestWindowTitleEditorExtension) ... ok
2022-09-25T13:12:38.6028972Z testMakeTitleWithCustomTemplate (tests.windowtitleeditor.TestWindowTitleEditorExtension) ... ok
2022-09-25T13:12:38.6697040Z testMonkeyPatch (tests.windowtitleeditor.TestWindowTitleEditorExtension) ... ok
2022-09-25T13:12:38.7409450Z testTeardown (tests.windowtitleeditor.TestWindowTitleEditorExtension) ... ok
2022-09-25T13:12:38.7527749Z 
2022-09-25T13:12:38.7528141Z ======================================================================
2022-09-25T13:12:38.7528676Z FAIL: testShowRecentChangesDialog (tests.uiactions.TestUIActions)
2022-09-25T13:12:38.7529441Z ----------------------------------------------------------------------
2022-09-25T13:12:38.7529824Z Traceback (most recent call last):
2022-09-25T13:12:38.7530447Z   File "D:/a\zim-desktop-wiki\zim-desktop-wiki/tests/uiactions.py", line 638, in testShowRecentChangesDialog
2022-09-25T13:12:38.7531045Z     self.assertEqual(self.navigation.opened, Path('NewPage'))
2022-09-25T13:12:38.7531481Z AssertionError: <Path: ExistingPage> != <Path: NewPage>
```

TL;DR imo ready to merge

**Comment by jaap-karssenberg** [comment]
Hi Introt,

Sorry, didn't have the time to merge it yesterday. Will put it on the list
for the next release - which I promise will be much faster than this one...
(Even didn't get to finish the release yesterday - the windows installer
build script is broken ...)

One change I would like to request (or I can make myself when merging) is
to not add a menu item in this plugin. Feels like clutter to me, since this
is really a preference that you set once and don't change often. So my
recommendation would be to just put the configuration in the plugin
preferences. Maybe you don't really need the templates either in that case,
just a string configuration for the custom template. But I'm fine with a
dropdown of default choices as well if you think that is better.

Kind regards,

Jaap

**Comment by introt** [comment]
Hi Jaap, and congrats on the new release!

No worries, the line had to be drawn somewhere.

I agree that the menu item is too much clutter, but I'd leave the build-in templates to make the feature more accessible to those who prefer the old title format - forcing them to read a manual page to overcome what they view as a usability regression feels unnecessary, but maybe I'm just a softie ;)

**Comment by introt** [comment]
I need a quick refresher on plugin references to make the changes, but I'd be happy to do it myself - it might be also possible to simplify some parts without the menu items. Are we talking about weeks or months for the next release?

**Comment by jaap-karssenberg** [comment]
To use preferences you need to define "plugin_preferences" in the plugin
object with the type and choices of each preference. In the extension
objects you can than access "self.plugin.preferences" as a dict with the
setting. There is a "changed" signal on the dict for updates.

See e.g. the backlinkpane plugin for a simple example. Here the pane
location is a multiple-choice dropdown.

My plan today is ~ 1 month untill the next dot release (0.75.1) with small
fixes & translations, ~ 3 months until 0.76 - but depending how fast I can
properly merge my shelved half broken patches ;)

Regards,

Jaap

**Comment by introt** [comment]
Thanks Jaap! I've made the changes, good luck with your patches!
